torchprime/torch_xla_models/tests/test_trainer_cpu.py

```diff
@@ -0,0 +1,174 @@
+
+from torchprime.metrics.metrics import MetricsLogger
+from torchprime.torch_xla_models.trainer.basic import Trainer
+
+
+class DummyModel(nn.Module):
+  def __init__(self):
+    super().__init__()
+    self.linear = nn.Linear(4, 2)
+
+  def forward(self, input_ids=None, attention_mask=None, **kwargs):
+    logits = self.linear(input_ids)
+    loss = logits.mean()
+    return logits, loss
+
+
+class DummyDataset(Dataset):
+  def __getitem__(self, idx):
+    return {"input_ids": torch.ones(4), "attention_mask": torch.ones(4)}
+
+  def __len__(self):
+    return 100
+
+
+class FakeMesh:
+  def __init__(self):
+    self.device_ids = [0]
+    self.axis_names = ("data", "fsdp")
+    self.mesh_shape = (1, 1)
+
+  def shape(self):
+    return {"data": 1, "fsdp": 1}
+
+  def get_axis_name_idx(self, axis_name):
+    return self.axis_names.index(axis_name)
+
+  def get_logical_mesh(self):
+    return np.array(self.device_ids).reshape(self.mesh_shape)
+
+
+@pytest.fixture
+def dummy_config():
+  return OmegaConf.create(
+    {
+      "global_batch_size": 4,
+      "max_steps": 2,
+      "output_dir": "/tmp/test_output",
+      "logging_steps": 1,
+      "profile_step": -1,
+      "profile_dir": "/tmp/profile",
+      "profile_duration": 5,
+      "optimizer": {"type": "adafactor", "learning_rate": 1e-3},
+      "lr_scheduler": {"type": "constant", "warmup_steps": 0},
+      "block_size": 4,
+      "model": {
+        "remat": {
+          "activation_checkpoint_layers": [],
+          "optimization_barrier_layers": [],
+          "scan_layers": None,
+          "offload_tensors": [],
+        },
+        "sharding": {"type": "spmd"},
+      },
+      "ici_mesh": {
+        "data": 1,
+        "fsdp": 1,
+        "tensor": 1,
+      },
+      "dcn_mesh": {},
+    }
+  )
+
+
+@patch(
+  "torchprime.torch_xla_models.sharding.initialization.get_mesh",
+  return_value=FakeMesh(),
+)
+@patch(
+  "torchprime.torch_xla_models.sharding.initialization.shard_torch_xla_model_from_config",
+  side_effect=lambda model, *args, **kwargs: model,
+)
+@patch("torchprime.torch_xla_models.trainer.basic.xm.xla_device", return_value="cpu")
+@patch("torchprime.torch_xla_models.trainer.basic.torch_xla.sync")
+def test_trainer_init(
+  mock_sync, mock_device, mock_shard_model, mock_get_mesh, dummy_config
+):
+  model = DummyModel()
+  dataset = DummyDataset()
+  trainer = Trainer(model, dummy_config, dataset)
+  assert isinstance(trainer.model, DummyModel)
+  assert trainer.global_batch_size == 4
+  assert trainer.device == "cpu"
+
+
+@patch(
+  "torchprime.torch_xla_models.sharding.initialization.get_mesh",
+  return_value=FakeMesh(),
+)
+@patch(
+  "torchprime.torch_xla_models.sharding.initialization.shard_torch_xla_model_from_config",
+  side_effect=lambda model, *args, **kwargs: model,
+)
+@patch("torchprime.torch_xla_models.trainer.basic.xm.xla_device", return_value="cpu")
+@patch("torchprime.torch_xla_models.trainer.basic.torch_xla.sync")
```

Comment:
Looks like you are mocking out all PyTorch/XLA specific logic, is there any reason to do so?
It limits the usefulness of such test, and this test is really useful, thanks for putting it together.
FYI, majority of tests in torchrpime is executed on CPU, without mocking any logic.
Would it be possible to make it working without mocking?
If not, please add a comment describing why you think such setup is not possible.

Reply:
I don't see why we shouldn't, this is a great test. Could you at least give it a try, and if doesn't work document problem